source/extensions/filters/network/dubbo_proxy/hessian_utils.cc

```diff
@@ -0,0 +1,534 @@
+#include "extensions/filters/network/dubbo_proxy/hessian_utils.h"
+
+#include <type_traits>
+
+#include "common/common/assert.h"
+#include "common/common/fmt.h"
+
+#include "extensions/filters/network/dubbo_proxy/buffer_helper.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace NetworkFilters {
+namespace DubboProxy {
+
+namespace {
+
+template <typename T>
+typename std::enable_if<std::is_signed<T>::value, T>::type leftShift(T left, uint16_t bit_number) {
+  if (left < 0) {
+    left = -left;
+    return -1 * (left << bit_number);
+  }
+
+  return left << bit_number;
+}
+
+} // namespace
+
+/*
+ * Reference:
+ * https://cs.chromium.org/chromium/src/base/strings/string_util.h?q=WriteInto&sq=package:chromium&dr=CSs&l=426
+ */
+char* allocStringBuffer(std::string* str, size_t length) {
+  str->reserve(length);
+  str->resize(length - 1);
+  return &((*str)[0]);
+}
+
+std::string HessianUtils::peekString(Buffer::Instance& buffer, size_t* size, uint64_t offset) {
+  ASSERT(buffer.length() > offset);
+  uint8_t code = buffer.peekInt<uint8_t>(offset);
+  size_t delta_length = 0;
+  std::string result;
+  switch (code) {
+  case 0x00:
+  case 0x01:
+  case 0x02:
+  case 0x03:
+  case 0x04:
+  case 0x05:
+  case 0x06:
+  case 0x07:
+  case 0x08:
+  case 0x09:
+  case 0x0a:
+  case 0x0b:
+  case 0x0c:
+  case 0x0d:
+  case 0x0e:
+  case 0x0f:
+  case 0x10:
+  case 0x11:
+  case 0x12:
+  case 0x13:
+  case 0x14:
+  case 0x15:
+  case 0x16:
+  case 0x17:
+  case 0x18:
+  case 0x19:
+  case 0x1a:
+  case 0x1b:
+  case 0x1c:
+  case 0x1d:
+  case 0x1e:
+  case 0x1f:
+    delta_length = code - 0x00;
+    if (delta_length + 1 + offset > buffer.length()) {
+      throw EnvoyException("buffer underflow");
+    }
+    buffer.copyOut(offset + 1, delta_length, allocStringBuffer(&result, delta_length));
+    *size = delta_length + 1;
+    return result;
+
+  case 0x30:
+  case 0x31:
+  case 0x32:
+  case 0x33:
+    if (offset + 2 > buffer.length()) {
+      throw EnvoyException("buffer underflow");
+    }
+
+    delta_length = (code - 0x30) * 256 + buffer.peekInt<int8_t>(offset + 1);
```

Comment:
peekInt<uint8_t>? Otherwise this would be underflow

Reply:
Here is the signed type hessian requirements, Hessian is written by Java, Only a signed types in Java

https://github.com/ebourg/hessian/blob/git-svn/src/com/caucho/hessian/io/Hessian2Input.java#L1183

Reply:
So `read()` there returns a `int` and with `& 0xFF` there it was converted to unsigned:
https://github.com/ebourg/hessian/blob/git-svn/src/com/caucho/hessian/io/Hessian2Input.java#L2820

https://stackoverflow.com/questions/4266756/can-we-make-unsigned-byte-in-java

Reply:
fixed